ssl/quic/quic_rstream.c

```diff
@@ -0,0 +1,143 @@
+/*
+* Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+*
+* Licensed under the Apache License 2.0 (the "License").  You may not use
+* this file except in compliance with the License.  You can obtain a copy
+* in the file LICENSE in the source distribution or at
+* https://www.openssl.org/source/license.html
+*/
+#include "internal/common.h"
+#include "internal/time.h"
+#include "internal/quic_stream.h"
+#include "internal/quic_sf_list.h"
+#include "internal/quic_fc.h"
+#include "internal/quic_error.h"
+
+struct quic_rstream_st {
+    SFRAME_LIST fl;
+    QUIC_RXFC *rxfc;
+    OSSL_STATM *statm;
+};
+
+QUIC_RSTREAM *ossl_quic_rstream_new(OSSL_QRX *qrx, QUIC_RXFC *rxfc,
+                                    OSSL_STATM *statm)
+{
+    QUIC_RSTREAM *ret = OPENSSL_malloc(sizeof(*ret));
+
+    if (ret == NULL)
+        return NULL;
+
+    ossl_sframe_list_init(&ret->fl, qrx);
+    ret->rxfc = rxfc;
+    ret->statm = statm;
+    return ret;
+}
+
+void ossl_quic_rstream_free(QUIC_RSTREAM *qrs)
+{
+    ossl_sframe_list_destroy(&qrs->fl);
+    OPENSSL_free(qrs);
+}
+
+int ossl_quic_rstream_queue_data(QUIC_RSTREAM *qrs, OSSL_QRX_PKT_WRAP *pkt_wrap,
+                                 uint64_t offset,
+                                 const unsigned char *data, uint64_t data_len,
+                                 int fin)
+{
+    UINT_RANGE range;
+
+    range.start = offset;
+    range.end = offset + data_len;
+
+    if (qrs->rxfc != NULL
+        && (!ossl_quic_rxfc_on_rx_stream_frame(qrs->rxfc, range.end, fin)
+            || ossl_quic_rxfc_get_error(qrs->rxfc, 0) != QUIC_ERR_NO_ERROR))
+        /* QUIC_ERR_FLOW_CONTROL_ERROR or QUIC_ERR_FINAL_SIZE detected */
+        return 0;
```

Comment:
I'll note it wouldn't be too unreasonable to make the RX depacketiser handle calling the RXFC; we could make `QUIC_RSTREAM` completely independent of the RXFC and decoupled from it.

Worth noting that as things stand, we can't distinguish between allocation failures in `ossl_sframe_list_insert` and a FC error. The CSM will probably need to make this distinction so it might be better to have this done in the depacketiser anyway.

An alternative would be multiple error codes on this API. I prefer that a little less, but I'd be OK with it.

Reply:
I agree, I'll do the refactoring.